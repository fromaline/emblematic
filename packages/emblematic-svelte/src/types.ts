import type { SvelteComponent } from 'svelte';
import type { SVGAttributes, SvelteHTMLElements } from 'svelte/elements';

export type Attrs = SVGAttributes<SVGSVGElement>;

export type IconNode = ([elementName: keyof SvelteHTMLElements, attrs: Attrs] | [elementName: keyof SvelteHTMLElements, attrs: Attrs, [elementName: keyof SvelteHTMLElements, attrs: Attrs][]])[];

export interface IconProps extends Attrs {
  size?: number | string;
  class?: string;
}

export type IconEvents = {
  [evt: string]: CustomEvent<any>;
};

export type IconSlots = {
  default: {};
};

export type Icon = SvelteComponent<IconProps, IconEvents, IconSlots>;