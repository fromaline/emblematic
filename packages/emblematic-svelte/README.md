# Emblematic Svelte

Implementation of the Emblematic icon library for svelte applications.
Emblematic is a collection of blockchain icons we use on [Chainspect](https://chainspect.app/).

## Installation

```sh
npm install emblematic-svelte
```

or

```sh
yarn add emblematic-svelte
```

or

```sh
pnpm add emblematic-svelte
```