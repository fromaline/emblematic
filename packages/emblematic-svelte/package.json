{
  "name": "emblematic-svelte",
  "description": "An Emblematic icon library package for Svelte applications",
  "version": "0.0.2",
  "bugs": {
    "url": "https://github.com/fromaline/emblematic/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/fromaline/emblematic.git",
    "directory": "packages/emblematic-svelte"
  },
  "keywords": [
    "Emblematic",
    "Svelte",
    "Icons",
    "Icon",
    "SVG",
    "Emblematic Icons"
  ],
  "author": "Nikita Grechino",
  "type": "module",
  "main": "dist/emblematic-svelte.js",
  "exports": {
    ".": {
      "types": "./dist/emblematic-svelte.d.ts",
      "svelte": "./dist/emblematic-svelte.js",
      "default": "./dist/emblematic-svelte.js"
    },
    "./icons": {
      "types": "./dist/emblematic-svelte.d.ts",
      "svelte": "./dist/emblematic-svelte.js"
    },
    "./icons/*": {
      "types": "./dist/icons/*.svelte.d.ts",
      "svelte": "./dist/icons/*.svelte"
    }
  },
  "typings": "dist/emblematic-svelte.d.ts",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "pnpm clean && pnpm copy:license && pnpm build:icons && pnpm build:package && pnpm build:license",
    "clean": "rm -rf dist && rm -rf stats && rm -rf ./src/icons/*.svelte && rm -f index.js",
    "copy:license": "cp ../../LICENSE ./LICENSE",
    "build:icons": "build-icons --output=./src --templateSrc=./scripts/exportTemplate.mjs --exportFileName=index.ts --iconFileExtension=.svelte --importImportFileExtension=.svelte --withAliases --aliasesFileExtension=.ts --aliasImportFileExtension=.svelte --pretty=false",
    "build:package": "svelte-package --input ./src",
    "build:license": "node ./scripts/appendBlockComments.mjs",
    "test": "vitest run"
  },
  "license": "MIT",
  "devDependencies": {
    "@emblematic/build-icons": "workspace:*",
    "@sveltejs/package": "^2.2.7",
    "@sveltejs/vite-plugin-svelte": "^3.0.2",
    "@tsconfig/svelte": "^5.0.2",
    "svelte": "^4.2.12",
    "typescript": "^5.3.3",
    "vite": "^5.1.4"
  },
  "peerDependencies": {
    "svelte": "^3 || ^4 || ^5.0.0-next.42"
  },
  "homepage": "https://github.com/fromaline/emblematic#readme"
}