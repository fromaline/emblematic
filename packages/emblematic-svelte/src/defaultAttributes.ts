import type { Attrs } from './types.js';

const defaultAttributes: Attrs = {
  xmlns: 'http://www.w3.org/2000/svg',
  width: 32,
  height: 32,
};

export default defaultAttributes;