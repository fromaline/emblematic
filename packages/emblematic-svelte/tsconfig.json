{
  "extends": "@tsconfig/svelte/tsconfig.json",
  "compilerOptions": {
    "target": "ESNext",
    "useDefineForClassFields": true,
    "module": "ESNext",
    "resolveJsonModule": true,
    "allowJs": true,
    "checkJs": true,
    "isolatedModules": true,
    "types": [],
  },
  "include": ["src/**/*.d.ts", "src/**/*.ts", "src/**/*.js", "src/**/*.svelte", "tests/**/*.ts"],
}