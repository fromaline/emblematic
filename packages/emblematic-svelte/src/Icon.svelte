<script lang="ts">
  import defaultAttributes from './defaultAttributes'
  import type { IconNode } from './types';

  export let name: string
  export let size: number | string = 24
  export let iconNode: IconNode
</script>

<svg
  {...defaultAttributes}
  {...$$restProps}
  width={size}
  height={size}
  class={`emblematic-icon emblematic emblematic-${name} ${$$props.class ?? ''}`}
>
  {#each iconNode as [tag, attrs]}
    <svelte:element this={tag} {...attrs}/>
  {/each}
  <slot />
</svg>