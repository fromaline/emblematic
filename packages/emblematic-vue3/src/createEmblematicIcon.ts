import { h } from 'vue'
import type { DefineComponent, FunctionalComponent, SVGAttributes } from 'vue'
import defaultAttributes from './defaultAttributes'

// Create interface extending SVGAttributes
export interface SVGProps extends Partial<SVGAttributes> {
  size?: 32 | number
}

export type IconNode = ([elementName: string, attrs: Record<string, string>] | [elementName: string, attrs: Record<string, string>, [elementName: string, attrs: Record<string, string>][]])[]
export type Icon = FunctionalComponent<SVGProps>
/**
 * Converts string to KebabCase
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export function toKebabCase(string: string) {
  return string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase()
}

function createEmblematicIcon(iconName: string, iconNode: IconNode): Icon {
  return (
    { size, class: _, ...props }, // props
    { attrs, slots }, // context
  ) => {
    return h(
      'svg',
      {
        ...defaultAttributes,
        width: size || defaultAttributes.width,
        height: size || defaultAttributes.height,
        ...attrs,
        class: ['emblematic', `emblematic-${toKebabCase(iconName)}`],
        ...props,
      },
      [...iconNode.map(([tag, attrs, children]) => {
        if (Array.isArray(children))
          return h(tag, attrs, children.map(([childTag, childAttrs]) => h(childTag, childAttrs)))
        else if (tag === 'title')
          return h(tag, attrs, attrs.values)

        return h(tag, attrs)
      }), ...(slots.default ? [slots.default()] : [])],
    )
  }
}

export default createEmblematicIcon
