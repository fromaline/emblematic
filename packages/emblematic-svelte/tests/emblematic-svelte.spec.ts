import { afterEach, describe, expect, it } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { Ethereum } from '../src/emblematic-svelte'
import TestSlots from './TestSlots.svelte'

describe('using Emblematic icon components', () => {
  afterEach(() => cleanup())

  it('should render an component', () => {
    const { container } = render(Ethereum)

    expect(container).toMatchSnapshot()
  })

  it('should adjust the size', () => {
    const { container } = render(Ethereum, {
      props: {
        size: 48,
      },
    })

    expect(container).toMatchSnapshot()
  })

  it('should add a class to the element', () => {
    const testClass = 'my-icon'
    const { container } = render(Ethereum, {
      props: {
        class: testClass,
      },
    })

    expect(container).toMatchSnapshot()

    const [el] = document.getElementsByClassName(testClass)

    expect(el).toBeInTheDocument()
    expect(el).toMatchSnapshot()
    expect(el).toHaveClass(testClass)
    expect(el).toHaveClass('emblematic')
    expect(el).toHaveClass('emblematic-ethereum')
  })

  it('should add a style attribute to the element', () => {
    const { container } = render(Ethereum, {
      props: {
        style: 'position: absolute;',
      },
    })

    expect(container).toMatchSnapshot()

    const [el] = document.getElementsByClassName('emblematic')

    expect(el.getAttribute('style')).toContain('position: absolute')
  })

  it('should render an icon slot', () => {
    const { container, getByText } = render(TestSlots)

    const textElement = getByText('Test')
    expect(textElement).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
