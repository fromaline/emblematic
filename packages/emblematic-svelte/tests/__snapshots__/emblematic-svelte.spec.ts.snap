// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`using Emblematic icon components > should add a class to the element 1`] = `
<body>
  <div>
    <svg
      class="emblematic-icon emblematic emblematic-ethereum my-icon"
      height="24"
      viewBox="0 0 32 32"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Ethereum
      </title>
      
      
      <path
        d="M16 0 6.2 16.3l9.8-4.5V0Z"
        fill="#8891B1"
      />
      
      
      <path
        d="m16 11.8-9.8 4.5 9.8 5.8V11.8Zm9.8 4.5L16 0v11.8l9.8 4.5Z"
        fill="#63688E"
      />
      
      
      <path
        d="m16 22.1 9.8-5.8-9.8-4.5v10.3Z"
        fill="#464A75"
      />
      
      
      <path
        d="M6.2 18.2 16 32v-8l-9.8-5.8Z"
        fill="#8891B1"
      />
      
      
      <path
        d="M16 24v8l9.8-13.8L16 24Z"
        fill="#63688E"
      />
      
      
      
    </svg>
  </div>
</body>
`;

exports[`using Emblematic icon components > should add a class to the element 2`] = `
<svg
  class="emblematic-icon emblematic emblematic-ethereum my-icon"
  height="24"
  viewBox="0 0 32 32"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <title>
    Ethereum
  </title>
  
  
  <path
    d="M16 0 6.2 16.3l9.8-4.5V0Z"
    fill="#8891B1"
  />
  
  
  <path
    d="m16 11.8-9.8 4.5 9.8 5.8V11.8Zm9.8 4.5L16 0v11.8l9.8 4.5Z"
    fill="#63688E"
  />
  
  
  <path
    d="m16 22.1 9.8-5.8-9.8-4.5v10.3Z"
    fill="#464A75"
  />
  
  
  <path
    d="M6.2 18.2 16 32v-8l-9.8-5.8Z"
    fill="#8891B1"
  />
  
  
  <path
    d="M16 24v8l9.8-13.8L16 24Z"
    fill="#63688E"
  />
  
  
  
</svg>
`;

exports[`using Emblematic icon components > should add a style attribute to the element 1`] = `
<body>
  <div>
    <svg
      class="emblematic-icon emblematic emblematic-ethereum "
      height="24"
      style="position: absolute;"
      viewBox="0 0 32 32"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Ethereum
      </title>
      
      
      <path
        d="M16 0 6.2 16.3l9.8-4.5V0Z"
        fill="#8891B1"
      />
      
      
      <path
        d="m16 11.8-9.8 4.5 9.8 5.8V11.8Zm9.8 4.5L16 0v11.8l9.8 4.5Z"
        fill="#63688E"
      />
      
      
      <path
        d="m16 22.1 9.8-5.8-9.8-4.5v10.3Z"
        fill="#464A75"
      />
      
      
      <path
        d="M6.2 18.2 16 32v-8l-9.8-5.8Z"
        fill="#8891B1"
      />
      
      
      <path
        d="M16 24v8l9.8-13.8L16 24Z"
        fill="#63688E"
      />
      
      
      
    </svg>
  </div>
</body>
`;

exports[`using Emblematic icon components > should adjust the size 1`] = `
<body>
  <div>
    <svg
      class="emblematic-icon emblematic emblematic-ethereum "
      height="48"
      viewBox="0 0 32 32"
      width="48"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Ethereum
      </title>
      
      
      <path
        d="M16 0 6.2 16.3l9.8-4.5V0Z"
        fill="#8891B1"
      />
      
      
      <path
        d="m16 11.8-9.8 4.5 9.8 5.8V11.8Zm9.8 4.5L16 0v11.8l9.8 4.5Z"
        fill="#63688E"
      />
      
      
      <path
        d="m16 22.1 9.8-5.8-9.8-4.5v10.3Z"
        fill="#464A75"
      />
      
      
      <path
        d="M6.2 18.2 16 32v-8l-9.8-5.8Z"
        fill="#8891B1"
      />
      
      
      <path
        d="M16 24v8l9.8-13.8L16 24Z"
        fill="#63688E"
      />
      
      
      
    </svg>
  </div>
</body>
`;

exports[`using Emblematic icon components > should render an component 1`] = `
<body>
  <div>
    <svg
      class="emblematic-icon emblematic emblematic-ethereum "
      height="24"
      viewBox="0 0 32 32"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Ethereum
      </title>
      
      
      <path
        d="M16 0 6.2 16.3l9.8-4.5V0Z"
        fill="#8891B1"
      />
      
      
      <path
        d="m16 11.8-9.8 4.5 9.8 5.8V11.8Zm9.8 4.5L16 0v11.8l9.8 4.5Z"
        fill="#63688E"
      />
      
      
      <path
        d="m16 22.1 9.8-5.8-9.8-4.5v10.3Z"
        fill="#464A75"
      />
      
      
      <path
        d="M6.2 18.2 16 32v-8l-9.8-5.8Z"
        fill="#8891B1"
      />
      
      
      <path
        d="M16 24v8l9.8-13.8L16 24Z"
        fill="#63688E"
      />
      
      
      
    </svg>
  </div>
</body>
`;

exports[`using Emblematic icon components > should render an icon slot 1`] = `
<body>
  <div>
    <svg
      class="emblematic-icon emblematic emblematic-algorand "
      height="24"
      viewBox="0 0 32 32"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Algorand
      </title>
      
      
      <path
        d="m8.16 28 3.49-6.01L15.13 16l3.46-6.01.57-.95.26.95 1.06 3.96L19.3 16l-3.48 5.99L12.35 28h4.16l3.48-6.01 1.8-3.11.86 3.1L24.26 28H28l-1.61-6.01L24.77 16l-.42-1.54 2.59-4.47h-3.78l-.13-.45-1.32-4.9-.17-.64h-3.63l-.08.13-3.4 5.86L10.95 16l-3.47 5.99L4 28h4.16Z"
        fill="#fff"
      />
      
      
      
      <text>
        Test
      </text>
    </svg>
  </div>
</body>
`;
